zebra-chain/src/orchard/keys.rs

```diff
@@ -705,6 +588,108 @@ impl PartialEq for FullViewingKey {
+/// §4.2.3][ps].
+///
+/// Used to decrypt incoming notes without spending them.
+///
+/// [ps]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+#[derive(Copy, Clone)]
+pub struct IncomingViewingKey {
+    dk: DiversifierKey,
+    // TODO: refine type
+    ivk: pallas::Scalar,
+}
+
+impl ConstantTimeEq for IncomingViewingKey {
+    /// Check whether two `IncomingViewingKey`s are equal, runtime independent
+    /// of the value of the secret.
+    ///
+    /// # Note
+    ///
+    /// This function should execute in time independent of the `dk` and `ivk` values.
+    fn ct_eq(&self, other: &Self) -> Choice {
+        <[u8; 64]>::from(*self).ct_eq(&<[u8; 64]>::from(*other))
+    }
+}
+
+impl fmt::Debug for IncomingViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("IncomingViewingKey")
+            .field("dk", &self.dk)
+            .field("ivk", &self.ivk)
+            .finish()
+    }
+}
+
+impl fmt::Display for IncomingViewingKey {
+    /// The _raw encoding_ of an **Orchard** _incoming viewing key_.
+    ///
+    /// <https://zips.z.cash/protocol/protocol.pdf#orchardfullviewingkeyencoding>
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&hex::encode(<[u8; 64]>::from(*self)))
+    }
+}
+
+impl Eq for IncomingViewingKey {}
+
+impl From<IncomingViewingKey> for [u8; 64] {
+    fn from(ivk: IncomingViewingKey) -> [u8; 64] {
+        let mut bytes = [0u8; 64];
+
+        bytes[..32].copy_from_slice(&<[u8; 32]>::from(ivk.dk));
+        bytes[32..].copy_from_slice(&<[u8; 32]>::from(ivk.ivk));
+
+        bytes
+    }
+}
+
+impl From<FullViewingKey> for IncomingViewingKey {
+    /// Commit^ivk_rivk(ak, nk) :=
+    ///     SinsemillaShortCommit_rcm(︁
+    ///        "z.cash:Orchard-CommitIvk",
+    ///        I2LEBSP_l^Orchard_base(ak) || I2LEBSP_l^Orchard_base(nk)︁
+    ///     ) mod r_P
+    ///
+    /// <https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents>
+    /// <https://zips.z.cash/protocol/nu5.pdf#concreteprfs>
+    #[allow(non_snake_case)]
+    fn from(fvk: FullViewingKey) -> Self {
+        let mut M: BitVec<Lsb0, u8> = BitVec::new();
+
+        // I2LEBSP_l^Orchard_base(ak)︁
+        let ak_bytes =
+            extract_p(pallas::Point::from_bytes(&fvk.spend_validating_key.into()).unwrap())
+                .to_bytes();
+        M.extend_from_bitslice(&BitArray::<Lsb0, _>::from(ak_bytes)[0..255]);
+
+        // I2LEBSP_l^Orchard_base(nk)︁
+        let nk_bytes: [u8; 32] = fvk.nullifier_deriving_key.into();
+        M.extend_from_bitslice(&BitArray::<Lsb0, _>::from(nk_bytes)[0..255]);
+
+        // Commit^ivk_rivk
+        // rivk needs to be 255 bits long
+        let commit_x = sinsemilla_short_commit(
+            fvk.ivk_commit_randomness.into(),
+            b"z.cash:Orchard-CommitIvk",
+            &M,
+        )
+        .expect("deriving orchard commit^ivk should not output ⊥ ");
```

Comment:
Question: I saw on the spec that if it's `⊥` (orthogonal?), then it should discard `ivk` and repeat with a new `sk`. Does `sinsemilla_short_commit` handle that?

(I'm fishing to try to learn something here :laughing:)

Reply:
Can we avoid unicode until all the common tools handle it correctly?

For example, `rust-analyzer` panics on unicode:
https://github.com/rust-analyzer/rust-analyzer/issues/9121

They implemented the [utf8 extension](https://clangd.llvm.org/extensions.html#utf-8-offsets
), but I'm still waiting for the emacs `lsp-mode` to activate it: https://github.com/emacs-lsp/lsp-mode/issues/2080

Reply:
To be clear, this isn't a blocker for this PR.

Unicode just makes it hard for me to work on the code, because `rust-analyzer` crashes after I view or edit parts of the code.

Reply:
As far as I can tell from those tickets, the problem should only occur for non-BMP characters (i.e. characters with code point > U+FFFF). If it happens for BMP characters, that would probably be useful to mention there. "⊥" is a BMP character, as are most mathematical symbols.

Reply:
Btw, ⊥ is defined in https://zips.z.cash/protocol/protocol.pdf#notation

Reply:
To clarify: `rust-analyzer` works with non-BMP codepoints (except maybe for that Zalgo issue which I can no longer reproduce), but panics when emacs feeds it bogus data.

Reply:
@jvff good question! `SpendingKey::new` handles the first failure case (ask == 0), but the second one you've pointed out seems to be missing. I've created a ticket for it in https://github.com/ZcashFoundation/zebra/issues/2496

@teor2345 there's a lot of Unicode in multiple files, so I'd prefer to do this in a separate PR, if we're going to do it

Reply:
> To clarify: `rust-analyzer` works with non-BMP codepoints (except maybe for that Zalgo issue which I can no longer reproduce), but panics when emacs feeds it bogus data.

To clarify even further:
`rust-analyzer` supports UTF-8 indexes, which emacs uses. But `emacs-lsp` doesn't set the "UTF-8 index" option, so
`rust-analyzer` assumes it is getting a UTF-16 index.

So ideally emacs would set the right option. But also `rust-analyzer` could just ignore the bogus index, or return an error.

Reply:
> @teor2345 there's a lot of Unicode in multiple files, so I'd prefer to do this in a separate PR, if we're going to do it

Let's see how it goes for now. If BMP unicode works, then we're mainly talking emoji, which we've removed already.